build/changelog/ChangelogBuilder.php

```diff
@@ -6,62 +6,72 @@
  */
 class ChangelogBuilder
 {
+    /** @var array */
+    private $params = [];
 
+    /** @var boolean */
     private $newServiceFlag = false;
-    private $baseDir;
-    private $releaseNotesOutputDir;
-    private $verboseFlag;
 
-    public function __construct($baseDir, $releaseNotesOutputDir, $verboseFlag)
+    /**
+     *  The constructor requires following configure parameters:
+     * - base_dir: (String) Path to the base directory where the `.changes` folder is located. Default is empty string.
+     * - release_notes_output_dir: (String) Path to `.changes` folder where new release json will be put and content of
+     *                              `nextrelease` folder will be deleted
+     * - verbose_flag: (Boolean) Flag to enable(true)/disable(false) verbose mode
+     */
+    public function __construct(array $params)
     {
-        $this->baseDir = $baseDir;
-        $this->releaseNotesOutputDir = $releaseNotesOutputDir;
-        $this->verboseFlag = $verboseFlag;
+        $this->params['base_dir'] = isset($params['base_dir']) ? $params['base_dir'] : '';
```

Comment:
Think of it this way: if there are no distinct advantages to using an internal parameters array, then why do it if you know it will prevent IDEs from autocompleting these fields? I personally think it's much clearer to see:
```php
class X {
    /**
     * @var string
     */
    private $property1;

    /**
     * @var int
     */
    private $property2;

    /**
     * @var bool
     */
    private $property3;
}
```

than it is to see:
```php
class X {
    /**
     * @var array
     */
    private $properties;
}
```

In the first case, you can explicitly call out the data type and name of each expected property, and IDEs will be able to warn on type mismatches. In the second case, you're forced to inspect the constructor to see what properties are available, and no type checking can be provided automatically. We accept community contributions to all PHP files in this repository (though build tool contributions are vanishingly rare), so we shouldn't complicate the developer experience unless there's a compelling reason to do so.

Additionally, automated tooling will flag dynamically declaring a new instance property as a code smell, but cannot do so if a method adds something new to the properties array. I just do not see any advantage to using an array of properties inside the class but can identify several disadvantages. It would also be a departure from standard practice in the SDK (e.g., [AwsClient](https://github.com/aws/aws-sdk-php/blob/master/src/AwsClient.php#L153)) which is why I'd like to understand the motivation.